crypto/param_build.c

```diff
@@ -0,0 +1,293 @@
+    memcpy(&pd->num, num, size);
+    return 1;
+}
+
+void ossl_param_bld_init(OSSL_PARAM_BLD *bld)
+{
+    memset(bld, 0, sizeof(*bld));
+}
+
+int ossl_param_bld_get_secure(const OSSL_PARAM_BLD *bld)
+{
+    return bld->is_secure;
+}
+
+int ossl_param_bld_set_secure(OSSL_PARAM_BLD *bld, int secure)
+{
+    int ret = bld->is_secure;
+
+    bld->is_secure = secure ? 1 : 0;
+    return ret;
+}
+
+int ossl_param_bld_push_int(OSSL_PARAM_BLD *bld, const char *key, int num)
+{
+    return param_push_num(bld, key, &num, sizeof(num), OSSL_PARAM_INTEGER);
+}
+
+int ossl_param_bld_push_uint(OSSL_PARAM_BLD *bld, const char *key,
+                             unsigned int num)
+{
+    return param_push_num(bld, key, &num, sizeof(num),
+                          OSSL_PARAM_UNSIGNED_INTEGER);
+}
+
+int ossl_param_bld_push_long(OSSL_PARAM_BLD *bld, const char *key,
+                             long int num)
+{
+    return param_push_num(bld, key, &num, sizeof(num), OSSL_PARAM_INTEGER);
+}
+
+int ossl_param_bld_push_ulong(OSSL_PARAM_BLD *bld, const char *key,
+                              unsigned long int num)
+{
+    return param_push_num(bld, key, &num, sizeof(num),
+                          OSSL_PARAM_UNSIGNED_INTEGER);
+}
+
+int ossl_param_bld_push_int32(OSSL_PARAM_BLD *bld, const char *key,
+                              int32_t num)
+{
+    return param_push_num(bld, key, &num, sizeof(num), OSSL_PARAM_INTEGER);
+}
+
+int ossl_param_bld_push_uint32(OSSL_PARAM_BLD *bld, const char *key,
+                               uint32_t num)
+{
+    return param_push_num(bld, key, &num, sizeof(num),
+                          OSSL_PARAM_UNSIGNED_INTEGER);
+}
+
+int ossl_param_bld_push_int64(OSSL_PARAM_BLD *bld, const char *key,
+                              int64_t num)
+{
+    return param_push_num(bld, key, &num, sizeof(num), OSSL_PARAM_INTEGER);
+}
+
+int ossl_param_bld_push_uint64(OSSL_PARAM_BLD *bld, const char *key,
+                               uint64_t num)
+{
+    return param_push_num(bld, key, &num, sizeof(num),
+                          OSSL_PARAM_UNSIGNED_INTEGER);
+}
+
+int ossl_param_bld_push_size_t(OSSL_PARAM_BLD *bld, const char *key,
+                               size_t num)
+{
+    return param_push_num(bld, key, &num, sizeof(num),
+                          OSSL_PARAM_UNSIGNED_INTEGER);
+}
+
+int ossl_param_bld_push_double(OSSL_PARAM_BLD *bld, const char *key,
+                               double num)
+{
+    return param_push_num(bld, key, &num, sizeof(num), OSSL_PARAM_REAL);
+}
+
+int ossl_param_bld_push_BN(OSSL_PARAM_BLD *bld, const char *key,
+                           const BIGNUM *bn)
+{
+    int sz = BN_num_bytes(bn);
+    OSSL_PARAM_BLD_DEF *pd;
+
+    if (sz <= 0) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_BLD_PUSH_BN, CRYPTO_R_ZERO_LENGTH_NUMBER);
+        return 0;
+    }
+    pd = param_push(bld, key, sz, sz, OSSL_PARAM_UNSIGNED_INTEGER);
+    if (pd == NULL)
+        return 0;
+    if (BN_get_flags(bn, BN_FLG_SECURE))
```

Comment:
so once you push a BN, the "secure" flag is sticky? if so, that seems really wrong.

Reply:
Once you push a _secure_ BN, values will be securely allocated.  That seems like the right thing to do.

The secure allocation can later be disabled by the caller but it requires an active step.

Reply:
so if i push a secure BN then a text string, then a non-secure BN what happens?
Suppose I push a keypair, and do the private-key part first?
Like I said https://github.com/openssl/openssl/pull/9305#discussion_r302026303, having a mode flag controlling such important semantics is bad.

Reply:
In both cases everything will be allocated in secure memory.

Reply:
>In both cases everything will be allocated in secure memory.

That is very bad.  the secure heap is a fixed-size resource, and wasting space by putting public key material there *is a bug*.  (In my view that makes this PR unacceptable, but I don't get a vote.)

Reply:
My initial reaction is to fix the secure heap so it isn't so restrictive.

Reply:
It is *still a bug in this code.*  It goes against all other users of the secure heap which only stored private key material, and did so by explicitly calling secure-new.

Reply:
Changed the code to have two allocation spaces.